### PR TITLE
Increase unit test coverage of pkg/cloud/aws/actuators/machine

### PR DESCRIPTION
**Pull request by ingvagabund**
Increase unit test coverage before:
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/actuator.go (62.1%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/awsclient.go (0.0%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/instances.go (52.9%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/loadbalancers.go (18.4%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/utils.go (53.9%)

Increase unit test coverage after:
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/actuator.go (91.6%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/awsclient.go (100.0%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/instances.go (91.2%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/loadbalancers.go (91.8%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/stubs.go (100.0%)
- sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine/utils.go (76.5%)

Goal:
- get coverage of each package above 90%

Current:
```sh
$ go test -race -cover ./pkg/cloud/aws/actuators/machine...
ok  	sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/machine	1.139s	coverage: 90.1% of statements
```

### CODE REVIEW

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 1]
```diff
@@ -34,6 +34,153 @@ func init() {
...
+					return mockAWSClient, nil
+				},
+				Codec: codec,
+				// use fake recorder and store an event into one item long buffer for subsequent check
+				EventRecorder: &record.FakeRecorder{
+					Events: eventsChannel,
+				},
+			}
+
+			if tc.error == launcInstanceError {
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: Could be reduced to:

    mockRunInstances(mockAWSClient, tc.error == launchInstanceError)

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 2]
```diff
@@ -34,6 +34,153 @@ func init() {
...
+				mockRunInstances(mockAWSClient, true)
+			} else {
+				mockRunInstances(mockAWSClient, false)
+			}
+
+			mockDescribeInstances(mockAWSClient, false)
+			mockTerminateInstances(mockAWSClient)
+
+			if tc.error == lbError {
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: Could be reduced to:

    mockRegisterInstancesWithLoadBalancer(mockAWSClient, <bool-expression>)

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 3]
```diff
@@ -181,6 +181,172 @@ func TestMachineEvents(t *testing.T) {
...
+	}
+
+	codec, err := providerconfigv1.NewCodec()
+	if err != nil {
+		t.Fatalf("unable to build codec: %v", err)
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: nit: drop blank line.

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 4]
```diff
@@ -195,38 +219,91 @@ func TestActuator(t *testing.T) {
...
 		terminateInstancesErr   error
 		lbErr                   error
 	}{
 		{
-			name:    "Update with success",
+			name:    "Create machine with success",
 			machine: machine,
-			operation: func(actuator *Actuator, cluster *clusterv1.Cluster, machine *clusterv1.Machine) {
+			operation: func(objectClient client.Client, actuator *Actuator, cluster *clusterv1.Cluster, machine *clusterv1.Machine) {
+				createErr := actuator.Create(context.TODO(), cluster, machine)
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: There's no checking of `createErr`. Can this fail?

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 5]
```diff
@@ -195,38 +219,91 @@ func TestActuator(t *testing.T) {
...
+				machineStatus, err := getMachineStatus(objectClient, machine)
+				if err != nil {
+					t.Fatalf("Unable to get machine status: %v", err)
+				}
+
+				assert.NoError(t, createErr)
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: Should this happen closer to the call to `actuator.Create()`?

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 6]
```diff
@@ -195,38 +219,91 @@ func TestActuator(t *testing.T) {
...
+				assert.Equal(t, machineStatus.Conditions[0].Reason, MachineCreationSucceeded)
+
+				// Get the machine
+				if exists, err := actuator.Exists(context.TODO(), cluster, machine); err != nil || !exists {
+					t.Errorf("Instance for %v does not exists: %v", strings.Join([]string{machine.Namespace, machine.Name}, "/"), err)
+				} else {
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: Isn't there a case here where there's no error, but it also doesn't exist?

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 7]
```diff
@@ -195,38 +219,91 @@ func TestActuator(t *testing.T) {
...
+					t.Logf("Instance for %v exists", strings.Join([]string{machine.Namespace, machine.Name}, "/"))
+				}
+
+				// Update a machine
+				if err := actuator.Update(context.TODO(), cluster, machine); err != nil {
+					t.Errorf("Unable to create instance for machine: %v", err)
+				}
+
+				// Get the machine
+				if exists, err := actuator.Exists(context.TODO(), cluster, machine); err != nil || !exists {
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: Isn't there a case here where there's no error, but it also doesn't exist?

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 8]
```diff
@@ -216,6 +230,15 @@ func TestActuator(t *testing.T) {
 	machineInvalidProviderConfig.Spec.ProviderConfig.Value = nil
 	machineInvalidProviderConfig.Spec.ProviderConfig.ValueFrom = nil
 
+	machineNoClusterID := machine.DeepCopy()
+	delete(machineNoClusterID.Labels, providerconfigv1.ClusterIDLabel)
+
+	pendingInstance := stubInstance("ami-a9acbbd6", "i-02fcb933c5da7085c")
+	pendingInstance.State = &ec2.InstanceState{
+		Name: aws.String(ec2.InstanceStateNamePending),
+		Code: aws.Int64(16),
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: Create a symbolic constant to convey what 16 means.

File: pkg/cloud/aws/actuators/machine/instaces_test.go [hunk 9]
```diff
@@ -84,3 +89,273 @@ func TestBuildEC2Filters(t *testing.T) {
...
+			mockAWSClient.EXPECT().TerminateInstances(gomock.Any()).AnyTimes()
+			removeStoppedMachine(machine, mockAWSClient)
+		})
+	}
+}
+
+func TestRunningInstace(t *testing.T) {
+	machine, _, err := stubMachine()
+	if err != nil {
+		t.Errorf("Unable to build test machine manifest: %v", err)
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: That will stop running test which fails

File: pkg/cloud/aws/actuators/machine/instaces_test.go [hunk 10]
```diff
@@ -84,3 +89,273 @@ func TestBuildEC2Filters(t *testing.T) {
...
+	mockAWSClient.EXPECT().DescribeInstances(gomock.Any()).Return(&ec2.DescribeInstancesOutput{
+		Reservations: []*ec2.Reservation{
+			{
+				Instances: []*ec2.Instance{
+					{
+						ImageId:    aws.String("ami-a9acbbd6"),
+						InstanceId: aws.String("i-02fcb933c5da7085c"),
+						State: &ec2.InstanceState{
+							Name: aws.String("Running"),
+							Code: aws.Int64(16),
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: Me neither, just copy paste. Though, I can remove that line.

File: pkg/cloud/aws/actuators/machine/instaces_test.go [hunk 11]
```diff
@@ -84,3 +89,273 @@ func TestBuildEC2Filters(t *testing.T) {
...
+		},
+		{
+			LaunchTime: nil,
+		},
+		{
+			LaunchTime: aws.Time(time.Now()),
+		},
+	}
+	sortInstances(instances)
+}
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: It just tests all the execution paths inside `sortInstances` are took

File: pkg/cloud/aws/actuators/machine/instaces_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]:
```console
func (*B) Errorf
func (c *B) Errorf(format string, args ...interface{})
Errorf is equivalent to Logf followed by Fail.

func (*B) Fail
func (c *B) Fail()
Fail marks the function as having failed but continues execution.
```

[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: `Errorf` will continue execution.

[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: If you run:

 ```go
func TestFoo(t *testing.T) {
	t.Errorf("this failed")
	t.Errorf("does this get printed")
}
```
"does this get printed" - will be printed. Execution continues in the curent TestFunction.

File: pkg/cloud/aws/actuators/machine/instaces_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: If the test doesn't require it, I would remove it.

File: pkg/cloud/aws/actuators/machine/stubs.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+		UserDataSecret: &corev1.LocalObjectReference{
+			Name: userDataSecretName,
+		},
+		Tags: []providerconfigv1.TagSpecification{
+			{Name: "openshift-node-group-config", Value: "node-config-master"},
+			{Name: "host-type", Value: "master"},
+			{Name: "sub-host-type", Value: "default"},
+		},
+		SecurityGroups: []providerconfigv1.AWSResourceReference{
+			{ID: aws.String("sg-00868b02fbe29de17")}, // aws-actuator
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/stubs.go]: Those `sg`'s no longer exist. Removing the comments.

File: pkg/cloud/aws/actuators/machine/stubs.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+			Namespace: defaultNamespace,
+		},
+		Data: map[string][]byte{
+			userDataSecretKey: []byte(userDataBlob),
+		},
+	}
+
+	machine, _, err := stubMachine()
+	if err != nil {
+		return nil, nil, nil, nil, err
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/stubs.go]: `stubMachineAPIResources()` should be removed eventually in favor of running `stubMachine()`, `stubCluster()`, `stubUserDataSecret()`, etc.

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 5 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: At the end there is no difference, the `machineStatus` needs to be retrievable from the machine object no matter of the `Create` operation failed or not.

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 6 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: The else branch is executed only when there is no error and the machine exist. See `err != nil || !exists` condition.

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 5 above)
Comment (frobware) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: If that's the case why not move it closer to the call. The pattern is call func, check for error in (really) most places. This stuck out, hence my comment/question.

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 7 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: The else branch is executed only when there is no error and the machine exist. See `err != nil || !exists` condition.

**Comment by ingvagabund** [comment]
@frobware comments addressed. PTAL

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 8 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Removed

File: pkg/cloud/aws/actuators/machine/instaces_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: Removed

File: pkg/cloud/aws/actuators/machine/instaces_test.go [hunk 14]
```diff
@@ -84,3 +89,273 @@ func TestBuildEC2Filters(t *testing.T) {
...
+			},
+		},
+	}, nil).AnyTimes()
+	getRunningInstance(machine, mockAWSClient)
+}
+
+func TestLaunchInstance(t *testing.T) {
+	machine, _, err := stubMachine()
+	if err != nil {
+		t.Errorf("Unable to build test machine manifest: %v", err)
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/instaces_test.go]: Replaced with Fatalf

File: pkg/cloud/aws/actuators/machine/awsclient_test.go [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+			output: &ec2.DescribeInstancesOutput{
+				Reservations: []*ec2.Reservation{
+					{
+						Instances: []*ec2.Instance{
+							{
+								ImageId:    aws.String("ami-a9acbbd6"),
+								InstanceId: aws.String("i-02fcb933c5da7085c"),
+								State: &ec2.InstanceState{
+									Name: aws.String("Running"),
+									Code: aws.Int64(16),
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/awsclient_test.go]: Removed

File: pkg/cloud/aws/actuators/machine/awsclient_test.go [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/mock/gomock"
+	mockaws "sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/client/mock"
+)
+
+func TestAwsClient(t *testing.T) {
+	machine, machinePC, err := stubMachine()
+	if err != nil {
+		t.Errorf("Unable to build test machine manifest: %v", err)
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/awsclient_test.go]: Replaced with Fatalf

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 1 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Done

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 2 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Done

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 3 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Done

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 5 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Done.

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 4 above)
Comment (ingvagabund) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Done. assert move right after the `Create` op call

**Pull request merged**